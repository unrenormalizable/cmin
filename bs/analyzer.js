export function analyze(program) {
  const warnings = []
  const errors = []

  let foundMain = false
  for (let i = 0; i < program.ast.functions.length; i++) {
    if (program.ast.functions[i].name === '__puts') {
      warnings.push({ token: program.ast.functions[i].token, message: 'redefining C imports.' })
    }
    if (program.ast.functions[i].name === 'main') {
      foundMain = true
    }
  }
  if (!foundMain) {
    errors.push({ location: { line: 0, column: 0 }, message: 'missing main function' })
  }

  return { warnings, errors }
}
