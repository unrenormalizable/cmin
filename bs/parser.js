import * as scanner from './scanner.js'

export const AST_NODE_PROGRAM = 0
export const AST_NODE_FUNCTION = 0

export const PARSE_ACCEPTED = 0
export const PARSE_NOT_YET_ACCEPTED = 1
export const PARSE_NOT_ACCEPTED = 2

function readToken(state, tokenId, tokenValue) {
  if (state.scanState.token.id === tokenId && (!tokenValue || state.scanState.token.value === tokenValue)) {
    const scanState = scanner.scan(state.text, state.scanState.state)
    state.scanState = scanState
    state.result = PARSE_NOT_YET_ACCEPTED
  } else {
    state.result = PARSE_NOT_ACCEPTED
  }

  return state
}

function parseFunction(state) {
  state = { ...state }

  state = readToken(state, scanner.TOK_KEYWORD, 'fn')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  const fnAst = { id: AST_NODE_FUNCTION, name: null, return: 'void', statements: [] }
  fnAst.name = state.scanState.token.value
  fnAst.token = state.scanState.token

  state = readToken(state, scanner.TOK_IDENTIFIER)
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, '(')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, ')')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, '{')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, '}')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state.ast.functions.push(fnAst)
  state.result = PARSE_NOT_YET_ACCEPTED
  return state
}

function parseFunctionList(state) {
  state = { ...state }

  if (state.scanState.token.id === scanner.TOK_EOF) {
    state.result = PARSE_NOT_YET_ACCEPTED
    return state
  }

  state = parseFunction(state)
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  return parseFunctionList(state)
}

function parseProgram(state) {
  state = { ...state }
  state.ast = { id: AST_NODE_PROGRAM, functions: [] }

  state = parseFunctionList(state)
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state.result = PARSE_ACCEPTED
  return state
}

export function parse(text) {
  const scanState = scanner.scan(text, { pos: 0, line: 1, column: 1 })
  const state = { text, result: PARSE_NOT_YET_ACCEPTED, scanState }

  return parseProgram(state)
}
