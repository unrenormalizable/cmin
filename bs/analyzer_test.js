import { assertSnapshot } from '@std/testing/snapshot'
import * as parser from './parser.js'
import * as analyzer from './analyzer.js'

Deno.test({
  name: 'analyze: override puts function',
  fn: async (tc) => {
    const text = 'fn __puts() {\n}'
    const ret = analyzer.analyze(parser.parse(text))

    await assertSnapshot(tc, ret, {})
  },
})
