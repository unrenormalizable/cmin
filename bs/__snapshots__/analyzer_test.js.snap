export const snapshot = {};

snapshot[`analyze: override puts function 1`] = `
{
  errors: [
    {
      location: {
        column: 0,
        line: 0,
      },
      message: "missing main function",
    },
  ],
  warnings: [
    {
      message: "redefining C imports.",
      token: {
        id: 4,
        location: {
          column: 4,
          line: 1,
          pos: 3,
        },
        value: "__puts",
      },
    },
  ],
}
`;
